File: iOS/Paging-CRUD/Paging-CRUD/ViewModel/CreatePostViewModel.swift
//
//  CreatePostViewModel.swift
//  Paging-CRUD
//

import Foundation

@MainActor
final class CreatePostViewModel: ObservableObject {
    @Published var state = State()

    private let repository: PostRepository

    init(repository: PostRepository = .shared) {
        self.repository = repository
    }

    private func insertPost(text: String) {
        Task {
            for await result in repository.addPost(apiKey: URLs.apiKey, groupId: Self.groupId, text: text) {
                switch result {
                case .success(let data):
                    state.textError = nil
                    state.isLoading = false
                    state.postId = data
                case .error(let message, _):
                    state.textError = nil
                    state.isLoading = false
                    state.message = message
                case .loading:
                    state.textError = nil
                    state.isLoading = true
                }
            }
        }
    }

    func actionSend() {
        if !state.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            insertPost(text: state.text)
        } else {
            state.textError = "내용을 입력해주세요."
        }
    }

    func onMessageShown() {
        state.message = ""
    }

    private static let groupId = 0

    struct State {
        var text = ""
        var textError: String? = nil
        var isLoading = false
        var postId = -1
        var message = ""
    }
}

File: Android/Paging-CRUD/app/src/main/java/com/hhp227/paging_crud/viewmodel/CreatePostViewModel.kt
package com.hhp227.paging_crud.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.paging_crud.data.PostRepository
import com.hhp227.paging_crud.model.Resource
import com.hhp227.paging_crud.util.URLs
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

class CreatePostViewModel internal constructor(
    private val repository: PostRepository
) : ViewModel() {
    val state = MutableStateFlow(State())

    private fun insertPost(text: String) {
        repository.addPost(URLs.API_KEY, GROUP_ID, text)
            .onEach { result ->
                when (result) {
                    is Resource.Success -> {
                        state.value = state.value.copy(
                            textError = null,
                            isLoading = false,
                            postId = result.data ?: -1
                        )
                    }
                    is Resource.Error -> {
                        state.value = state.value.copy(
                            textError = null,
                            isLoading = false,
                            message = result.message ?: "An unexpected error occured"
                        )
                    }
                    is Resource.Loading -> {
                        state.value = state.value.copy(textError = null, isLoading = true)
                    }
                }
            }
            .launchIn(viewModelScope)
    }

    fun onTextChange(text: String) {
        state.value = state.value.copy(text = text, textError = null)
    }

    fun actionSend() {
        if (state.value.text.isNotBlank()) {
            insertPost(state.value.text)
        } else {
            state.value = state.value.copy(textError = "내용을 입력해주세요.")
        }
    }

    fun onMessageShown() {
        state.value = state.value.copy(message = "")
    }

    companion object {
        private const val GROUP_ID = 0
    }

    data class State(
        val text: String = "",
        val textError: String? = null,
        val isLoading: Boolean = false,
        val postId: Int = -1,
        val message: String = ""
    )
}
